<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Case study 3: a multi-strain pathogen system with cross-reactive antibody dynamics • serosim</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Case study 3: a multi-strain pathogen system with cross-reactive antibody dynamics">
<meta property="og:description" content="serosim">
<meta property="og:image" content="/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">serosim</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/case_study_1_measles.html">Case study 1: a one pathogen system with vaccination (measles)</a>
    </li>
    <li>
      <a href="../articles/case_study_2_diphtheria_pertussis.html">Case study 2: a multi-pathogen system with multivalent vaccinaton (diphtheria and pertussis)</a>
    </li>
    <li>
      <a href="../articles/case_study_3_influenza.html">Case study 3: a multi-strain pathogen system with cross-reactive antibody dynamics</a>
    </li>
    <li>
      <a href="../articles/case_study_4_inference.html">Case study 4: Using serosim to assess the performance of seroepidemiological inference methods</a>
    </li>
    <li>
      <a href="../articles/quickstart.html">quickstart: Extended version of the README simulation</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/seroanalytics/serosim/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Case study 3: a multi-strain pathogen system
with cross-reactive antibody dynamics</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/seroanalytics/serosim/blob/HEAD/vignettes/case_study_3_influenza.Rmd" class="external-link"><code>vignettes/case_study_3_influenza.Rmd</code></a></small>
      <div class="hidden name"><code>case_study_3_influenza.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="outline">1. Outline<a class="anchor" aria-label="anchor" href="#outline"></a>
</h2>
<p>This vignette demonstrates how <em>serosim</em> can simulate complex
antibody profiles arising from repeated exposure to antigenically
related strains of the same pathogen, with a case study based on A/H3N2
influenza. It is recommended that you read the quick-start guide before
working through this case study. In this system, each exposure
stimulates a combination of <em>de novo</em> and memory antibody
responses, resulting in the presence of antibodies effective against all
strains. The effectiveness of these antibodies against a particular
strain depends on the set and timing of strains the individual has been
exposed to, and the antigenic similarity of those strains <span class="citation">Krammer (2019)</span>. The simulation will capture this
process by tracking infection dynamics and antibody profiles to multiple
strains for 100 individuals across a 45 year period, demonstrating how
complex antibody profiles, or antibody landscapes, build over the life
course <span class="citation">Lessler et al. (2012)</span>. A key
concept for this simulation is antigenic space – a conceptual framework
for quantifying the antigenic relatedness of strains from the
perspective of the immune response <span class="citation">(Smith et al.
2004)</span>. We will simulate a longitudinal serological survey with
discretized paired measurements (based on the haemagglutination
inhibition assay) of each individual’s antibody levels against each
strain based on existing serosurvey data <span class="citation">(Yang et
al. 2020)</span>.</p>
<p>Within each section below, we will briefly explain the rationale
behind the model structure and parameters. We caution users to conduct
their own research into the models and associated parameters which best
align with their disease system and biomarker test kits.</p>
<p>Load necessary packages:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Load serosim </span></span>
<span><span class="co">## devtools::install_github("seroanalytics/serosim")</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/seroanalytics/serosim" class="external-link">serosim</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Load additional packages required </span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://sjmgarnier.github.io/viridis/" class="external-link">viridis</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="ah3n2-influenza-epidemiology-and-simulation-settings">2. A/H3N2 influenza epidemiology and simulation settings<a class="anchor" aria-label="anchor" href="#ah3n2-influenza-epidemiology-and-simulation-settings"></a>
</h2>
<p>A/H3N2 influenza has distinctive epidemiology, whereby one group of
antigenically similar viruses (termed a cluster) circulates for 3-5
years before being replaced by a new, antigenically distinct cluster
<span class="citation">(Smith et al. 2004)</span>. The phylogeny of
A/H3N2 viruses is often referred to as ladder-like due to the fact that
influenza demonstrates limited genetic diversity at a single point in
time, but substantial diversity over the long-term. We will capture this
in our simulation by assuming that 10 antigenically distinct strains
each circulate for 4.5 years before being replaced by the next strain,
roughly representing the introduction of A/H3N2 into humans in 1968. The
simulation will use monthly time steps to balance computational time
with sufficient time-resolution to represent transient antibody
kinetics.</p>
<p>For this simulation, the only demography variable we will track is
age – individuals cannot be exposed to influenza strains which
circulated before they were born.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Specify the number of time periods to simulate </span></span>
<span><span class="va">times</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">4.5</span><span class="op">*</span><span class="fl">10</span><span class="op">*</span><span class="fl">12</span>,by<span class="op">=</span><span class="fl">1</span><span class="op">)</span> </span>
<span><span class="va">simulation_settings</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"t_start"</span><span class="op">=</span><span class="fl">1</span>,<span class="st">"t_end"</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">times</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Generate the population demography tibble</span></span>
<span><span class="va">N</span><span class="op">&lt;-</span><span class="fl">100</span></span>
<span><span class="co">## See help file(?generate_pop_demography) for more information on function arguments.</span></span>
<span><span class="co">## age_min is set to 0 which allows births to occur until the last time step</span></span>
<span><span class="co">## Let's assume that no individuals are removed from the population and set prob_removal to 0</span></span>
<span><span class="va">demography</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/generate_pop_demography.html">generate_pop_demography</a></span><span class="op">(</span><span class="va">N</span>, <span class="va">times</span>, age_min<span class="op">=</span><span class="fl">0</span>, prob_removal<span class="op">=</span><span class="fl">0</span><span class="op">)</span></span>
<span></span>
<span><span class="va">n_strains</span> <span class="op">&lt;-</span> <span class="fl">10</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="exposure-to-biomarker-mapping">3. Exposure to biomarker mapping<a class="anchor" aria-label="anchor" href="#exposure-to-biomarker-mapping"></a>
</h2>
<p>One of the key features of influenza serology is that different
strains of the same subtype are antigenically related due to the
presence of both conserved and distinct epitopes on surface proteins of
the strains. Strain replacements are the result of mutations at key
epitopes which allow for immune escape. Thus, an individual’s immune
profile is the culmination of antibodies targeting combinations of
epitopes consistent with the set of strains they have been exposed
to.</p>
<p>There are two ways we could model these dynamics with
<em>serosim</em>. First, we could map each strain to a set of epitopes,
such that sets of epitopes partially overlap between strains. The
exposure ID variable would correspond to each strain, and the biomarker
IDs would correspond to antibodies binding to each specific epitope.
Real data at this resolution are rare, particularly for
serosurveillance, requiring a complex assays (such as phage
immunoprecipitation sequencing, PhIP-Seq) which detects antibody binding
to short peptide sequences. The data demands and computational
complexity make this as a challenging model to simulate.</p>
<p>The second approach, and the one we will take here, is to consider
each biomarker ID as the output of a common assay, such as a
microneutralization assay or haemagglutination inhibition assay. These
assays measure the overall binding or reactivity of antibodies in a
serum sample against whole proteins or virions, and thus represent the
aggregate effect of all epitope-level binding. A high value on this sort
of assay reflects the joint effect of high antibody levels against
epitopes present in the assay; a lower value reflects lower antibody
concentrations and/or recognition of fewer epitopes. Strains are
considered to cross-react when antibodies elicited following exposure to
one strain produce a measurable effect against another strain,
reflecting the sharing of epitopes between the two strains.</p>
<p>For the <em>serosim</em> biomarker map, we model this system as an
all-to-all mapping between each strain an individual could be exposed to
(exposure ID) and each strain that could be included in the assay
(biomarker ID).</p>
<p>Another key feature of influenza epidemiology is antigenic drift,
where the accumulation of amino acid substitutions in surface proteins
are selected for by population immunity leading to a gradual change in
the virus phenotype over time. For this simulation, we will represent
antigenic drift as a simple random walk with log-normally distributed
moves through 2-dimensional antigenic space. We then assume a simple
cross-reactivity model, where the amount of cross-reactivity elicited
from infection with a particular strain to another strain is a function
of their euclidean distance in this space.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Create a matrix giving cross-reactivity between each strain pairing</span></span>
<span><span class="va">x_coord</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">numeric</a></span><span class="op">(</span><span class="va">n_strains</span><span class="op">)</span></span>
<span><span class="va">y_coord</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">numeric</a></span><span class="op">(</span><span class="va">n_strains</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Simulate antigenic drift as a random walk in two dimensions</span></span>
<span><span class="va">x_coord</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">0</span></span>
<span><span class="va">y_coord</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">0</span></span>
<span><span class="co">## Each strain is antigenically drifted from the previous one with log-normally distributed </span></span>
<span><span class="co">## moves through 2-dimensional antigenic space</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">2</span><span class="op">:</span><span class="va">n_strains</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">x_coord</span><span class="op">[</span><span class="va">i</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">x_coord</span><span class="op">[</span><span class="va">i</span><span class="op">-</span><span class="fl">1</span><span class="op">]</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/r/stats/Lognormal.html" class="external-link">rlnorm</a></span><span class="op">(</span><span class="fl">1</span>, mean<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span>, sd<span class="op">=</span><span class="fl">0.5</span><span class="op">)</span></span>
<span>  <span class="va">y_coord</span><span class="op">[</span><span class="va">i</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">y_coord</span><span class="op">[</span><span class="va">i</span><span class="op">-</span><span class="fl">1</span><span class="op">]</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/r/stats/Lognormal.html" class="external-link">rlnorm</a></span><span class="op">(</span><span class="fl">1</span>, mean<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span>, sd<span class="op">=</span><span class="fl">0.5</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co">## Cross-reactivity is a function of antigenic distance of the form e^-b*x, where x is the </span></span>
<span><span class="co">## euclidean distance between strains and b is a parameter</span></span>
<span><span class="va">antigenic_map</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>x_coord<span class="op">=</span><span class="va">x_coord</span>, y_coord<span class="op">=</span><span class="va">y_coord</span><span class="op">)</span></span>
<span><span class="va">antigenic_map</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/dist.html" class="external-link">dist</a></span><span class="op">(</span><span class="va">antigenic_map</span>, diag<span class="op">=</span><span class="cn">TRUE</span>, upper<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">##############</span></span>
<span><span class="va">antigenic_map</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">exp</a></span><span class="op">(</span><span class="op">-</span><span class="fl">0.25</span><span class="op">*</span><span class="va">antigenic_map</span><span class="op">)</span></span>
<span><span class="co">##############</span></span>
<span></span>
<span><span class="va">antigenic_map</span> <span class="op">&lt;-</span> <span class="va">antigenic_map</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>exposure_id<span class="op">=</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/context.html" class="external-link">n</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html" class="external-link">pivot_longer</a></span><span class="op">(</span><span class="op">-</span><span class="va">exposure_id</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html" class="external-link">rename</a></span><span class="op">(</span>biomarker_id<span class="op">=</span><span class="va">name</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>biomarker_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">biomarker_id</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="exposure-model">4. Exposure model<a class="anchor" aria-label="anchor" href="#exposure-model"></a>
</h2>
<p>We now turn to the exposure model determining the probability that an
individual is exposed to each strain at each time point. It is
straightforward to represent the serial replacement of strains in
<em>serosim</em> by setting the exposure ID dimension in the force of
exposure array to be zero at times before and after the strain
circulates. Furthermore, we can add seasonality to the model by
multiplying the force of exposure by a sinusoidal term <span class="citation">(<strong>Grassly2006?</strong>)</span>. We will set the
force of exposure to have annual cycles with a peak basic reproductive
number of 2, assuming a 5-day infectious period, though note that the
exposure model simply describes the force of exposure acting on
individuals in the simulation, and is not a dynamical transmission
model.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Seasonal forcing with a peak force of infection of 0.2 and minimum of 0</span></span>
<span><span class="co">## If we assume an infectious period of 5 days, the simulation gives an annually oscillating R0 between 0 and 2 </span></span>
<span><span class="va">beta</span> <span class="op">&lt;-</span> <span class="fl">0.2</span></span>
<span><span class="va">sigma</span> <span class="op">&lt;-</span> <span class="fl">1</span></span>
<span><span class="va">seasonality</span> <span class="op">&lt;-</span> <span class="va">beta</span><span class="op">*</span><span class="op">(</span><span class="fl">1</span> <span class="op">+</span> <span class="va">sigma</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/base/Trig.html" class="external-link">cos</a></span><span class="op">(</span><span class="fl">2</span><span class="op">*</span><span class="va">pi</span><span class="op">*</span><span class="op">(</span><span class="va">times</span><span class="op">-</span><span class="fl">1</span><span class="op">)</span><span class="op">/</span><span class="fl">12</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Create an empty array to store the force of exposure for all exposure types</span></span>
<span><span class="va">foe_pars</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/array.html" class="external-link">array</a></span><span class="op">(</span><span class="fl">0</span>, dim<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">times</span><span class="op">)</span>,<span class="va">n_strains</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">t_start</span> <span class="op">&lt;-</span> <span class="fl">1</span></span>
<span></span>
<span><span class="co">## Set the circulation duration for each strain</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">strain</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="va">n_strains</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">t_end</span> <span class="op">&lt;-</span> <span class="va">t_start</span> <span class="op">+</span> <span class="fl">4.5</span><span class="op">*</span><span class="fl">12</span> <span class="op">-</span> <span class="fl">1</span></span>
<span>  <span class="va">foe_pars</span><span class="op">[</span>,<span class="va">t_start</span><span class="op">:</span><span class="va">t_end</span>,<span class="va">strain</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">seasonality</span><span class="op">[</span><span class="va">t_start</span><span class="op">:</span><span class="va">t_end</span><span class="op">]</span></span>
<span>  <span class="va">t_start</span> <span class="op">&lt;-</span> <span class="va">t_end</span> <span class="op">+</span> <span class="fl">1</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co">## Specify a simple exposure model which calculates the probability of exposure directly from the force of exposure at that time step. In this selected model, the probability of exposure is 1-exp(-FOE) where FOE is the force of exposure at that time.</span></span>
<span><span class="va">exposure_model</span><span class="op">&lt;-</span><span class="va">exposure_model_simple_FOE</span></span></code></pre></div>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Examine the probability of exposure to each strain over time for the specified exposure model and foe_pars array</span></span>
<span><span class="fu"><a href="../reference/plot_exposure_model.html">plot_exposure_model</a></span><span class="op">(</span>exposure_model<span class="op">=</span><span class="va">exposure_model_simple_FOE</span>, times<span class="op">=</span><span class="va">times</span>, n_groups <span class="op">=</span> <span class="fl">1</span>, n_exposures <span class="op">=</span> <span class="fl">10</span>, foe_pars<span class="op">=</span><span class="va">foe_pars</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">exposure_id</span><span class="op">)</span></span></code></pre></div>
<p><img src="case_study_3_influenza_files/figure-html/unnamed-chunk-6-1.png" width="576"></p>
</div>
<div class="section level2">
<h2 id="immunity-model">5. Immunity model<a class="anchor" aria-label="anchor" href="#immunity-model"></a>
</h2>
<p>The immunity model determines the probability that an exposure event
induces an immunological response and becomes recorded in the immune
history. For influenza A/H3N2, a common correlate of protection is
haemagglutination inhibition assay (HI) titre against the exposure
strain. Human challenge studies can be used to parameterize a logistic
model for the relationship between HI titre and probability of
infection, which forms the basis for the model we will use here <span class="citation">(<strong>Coudeville2010?</strong>)</span>.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Specify immunity model within the runserosim function below </span></span>
<span><span class="va">immunity_model</span><span class="op">&lt;-</span><span class="va">immunity_model_ifxn_biomarker_prot</span></span>
<span></span>
<span><span class="co">## Define antibody-mediated protection parameters</span></span>
<span><span class="va">model_pars_immunity</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span>name<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"biomarker_prot_midpoint"</span>,<span class="st">"biomarker_prot_width"</span><span class="op">)</span>,</span>
<span>               mean<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2.8</span>,<span class="fl">1.3</span><span class="op">)</span>,sd<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="cn">NA</span>,<span class="cn">NA</span><span class="op">)</span>,distribution<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="cn">NA</span>,<span class="cn">NA</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">model_pars_immunity</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tidyr.tidyverse.org/reference/expand_grid.html" class="external-link">expand_grid</a></span><span class="op">(</span>biomarker_id<span class="op">=</span><span class="fl">1</span><span class="op">:</span><span class="va">n_strains</span>, <span class="va">model_pars_immunity</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>exposure_id<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Strain_"</span>,<span class="va">biomarker_id</span><span class="op">)</span>, biomarker_id<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Biomarker_"</span>,<span class="va">biomarker_id</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Parameters chosen from Coudeville et al. 2010</span></span>
<span><span class="fu"><a href="../reference/plot_biomarker_mediated_protection.html">plot_biomarker_mediated_protection</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">8</span>,by<span class="op">=</span><span class="fl">0.1</span><span class="op">)</span>, biomarker_prot_midpoint<span class="op">=</span><span class="fl">2.8</span>, biomarker_prot_width<span class="op">=</span><span class="fl">1.3</span><span class="op">)</span> </span></code></pre></div>
<p><img src="case_study_3_influenza_files/figure-html/unnamed-chunk-7-1.png" width="480"></p>
</div>
<div class="section level2">
<h2 id="antibody-model-and-parameters">6. Antibody model and parameters<a class="anchor" aria-label="anchor" href="#antibody-model-and-parameters"></a>
</h2>
<p>Models of antibody kinetics following A/H3N2 influenza infection vary
in their complexity, and identifying key immunological mechanisms of
waning, imprinting and back-boosting remains an area of ongoing research
<span class="citation">(<strong>Ranjeva2019?</strong>)</span>. For this
simulation, we assume a simple monophasic boosting and waning model,
where individuals receive a boost of approximately 4 HI titre units
following infection against the biomarker ID of the exposure strain
which wanes slowly over time. Cross-reactive boosting to other biomarker
IDs is determined by antigenic distance as described above.</p>
<p>To represent heterogeneity in post-infection antibody kinetics, we
will assume that individual boosting and waning parameters for each
infection event are log-normally distributed with mean and standard
deviation specified on the natural scale.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Specify the antibody model </span></span>
<span><span class="va">antibody_model</span><span class="op">&lt;-</span><span class="va">antibody_model_monophasic_cross_reactivity</span></span>
<span></span>
<span><span class="co">## Antibody model parameters</span></span>
<span><span class="va">model_pars_kinetics</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span>name<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"boost"</span>,<span class="st">"wane"</span><span class="op">)</span>,mean<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">4</span>,<span class="fl">0.01</span><span class="op">)</span>,sd<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">0.01</span><span class="op">)</span>,distribution<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"log-normal"</span>,<span class="st">"log-normal"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">model_pars_kinetics</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tidyr.tidyverse.org/reference/expand_grid.html" class="external-link">expand_grid</a></span><span class="op">(</span>exposure_id <span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="va">n_strains</span><span class="op">)</span>,<span class="va">model_pars_kinetics</span><span class="op">)</span>  <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>biomarker_id<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Biomarker_"</span>, <span class="va">exposure_id</span><span class="op">)</span>, exposure_id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Strain_"</span>,<span class="va">exposure_id</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">model_pars_observation</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tidyr.tidyverse.org/reference/expand_grid.html" class="external-link">expand_grid</a></span><span class="op">(</span>biomarker_id<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Biomarker_"</span>,<span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="va">n_strains</span><span class="op">)</span><span class="op">)</span>,name<span class="op">=</span><span class="st">"obs_sd"</span>,</span>
<span>                                      mean<span class="op">=</span><span class="cn">NA</span>,sd<span class="op">=</span><span class="fl">1</span>,distribution<span class="op">=</span><span class="st">"normal"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>exposure_id<span class="op">=</span><span class="cn">NA</span><span class="op">)</span></span>
<span><span class="va">model_pars_original</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind_rows.html" class="external-link">bind_rows</a></span><span class="op">(</span><span class="va">model_pars_kinetics</span>,<span class="va">model_pars_immunity</span>, <span class="va">model_pars_observation</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="va">exposure_id</span>, <span class="va">biomarker_id</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Bring in the antibody parameters needed for the antibody model</span></span>
<span><span class="co">## Note that the observation error parameter needed for the observation model (Section 7) is defined here too.</span></span>
<span></span>
<span><span class="co">## Reformat model_pars for runserosim</span></span>
<span><span class="va">model_pars</span><span class="op">&lt;-</span><span class="fu"><a href="../reference/reformat_biomarker_map.html">reformat_biomarker_map</a></span><span class="op">(</span><span class="va">model_pars_original</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">model_pars</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 6 × 6</span></span></span>
<span><span class="co">##   exposure_id name                     mean    sd distribution biomarker_id</span></span>
<span><span class="co">##         <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>               <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span>           1 boost                    4     1    log-normal              1</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span>           1 wane                     0.01  0.01 log-normal              1</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">3</span>           1 biomarker_prot_midpoint  2.8  <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>                      1</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">4</span>           1 biomarker_prot_width     1.3  <span style="color: #BB0000;">NA</span>    <span style="color: #BB0000;">NA</span>                      1</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">5</span>           2 boost                    4     1    log-normal              2</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">6</span>           2 wane                     0.01  0.01 log-normal              2</span></span></code></pre>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Specify the draw_parameters function</span></span>
<span><span class="va">draw_parameters</span><span class="op">&lt;-</span><span class="va">draw_parameters_random_fx</span></span></code></pre></div>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Plot example biomarker trajectories given the specified antibody kinetics model, model parameters and draw parameters function </span></span>
<span><span class="fu"><a href="../reference/plot_antibody_model.html">plot_antibody_model</a></span><span class="op">(</span><span class="va">antibody_model_monophasic_cross_reactivity</span>, N<span class="op">=</span><span class="fl">10</span>, model_pars<span class="op">=</span><span class="va">model_pars</span>,draw_parameters_fn <span class="op">=</span> <span class="va">draw_parameters_random_fx</span>, biomarker_map<span class="op">=</span><span class="va">antigenic_map</span><span class="op">)</span></span></code></pre></div>
<p><img src="case_study_3_influenza_files/figure-html/unnamed-chunk-9-1.png" width="960"></p>
</div>
<div class="section level2">
<h2 id="observation-model-and-sampling-times">7. Observation model and sampling times<a class="anchor" aria-label="anchor" href="#observation-model-and-sampling-times"></a>
</h2>
<p>Finally, we implement the observation model to reflect the
haemagglutination inhibition assay which is often used for influenza
serology <span class="citation">(<strong>Fonville2014?</strong>)</span>.
This is a discretized assay with some measurement error, which we will
represent using the <code>observation_model_discrete_noise</code>
function. The assay is bounded between values of 0 and 8. We will assume
high assay sensitivity and specificity, noting that many true positive
samples will be simulated with a HI titre of 0 through the noise
function.</p>
<p>To simulate a longitudinal survey with paired samples, we simulate
two observation times for each individual centered either around the
last time step or one year prior.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Specify the observation model </span></span>
<span><span class="va">observation_model</span><span class="op">&lt;-</span><span class="va">observation_model_discrete_noise</span></span>
<span></span>
<span><span class="co">## Specify the cutoffs for the discrete assay. This will depend on the dilutions of the haemagglutination assay.</span></span>
<span><span class="co">## This is a matrix with each row containing all of the cutoffs for that biomarker. Here, we have set the same cutoffs for all biomarkers </span></span>
<span><span class="va">breaks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">8</span>,by<span class="op">=</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">cutoffs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="va">breaks</span>,nrow<span class="op">=</span><span class="va">n_strains</span>,ncol<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">breaks</span><span class="op">)</span>, byrow<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Specify assay sensitivity and specificity needed for the observation model</span></span>
<span><span class="va">sensitivity</span><span class="op">&lt;-</span><span class="fl">0.95</span></span>
<span><span class="va">specificity</span><span class="op">&lt;-</span><span class="fl">0.99</span></span>
<span></span>
<span><span class="co">## Specify observation_times (serological survey sampling design) to observe all biomarkers across all individuals at the end of the simulation</span></span>
<span><span class="va">obs1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tidyr.tidyverse.org/reference/expand_grid.html" class="external-link">expand_grid</a></span><span class="op">(</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span>i<span class="op">=</span><span class="fl">1</span><span class="op">:</span><span class="va">N</span>, t <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">floor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="va">N</span>, <span class="fl">528</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>, b<span class="op">=</span><span class="fl">1</span><span class="op">:</span><span class="va">n_strains</span><span class="op">)</span></span>
<span><span class="va">obs2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tidyr.tidyverse.org/reference/expand_grid.html" class="external-link">expand_grid</a></span><span class="op">(</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span>i<span class="op">=</span><span class="fl">1</span><span class="op">:</span><span class="va">N</span>, t <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">pmin</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">floor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="va">N</span>, <span class="fl">540</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span>, <span class="fl">540</span><span class="op">)</span><span class="op">)</span>, b<span class="op">=</span><span class="fl">1</span><span class="op">:</span><span class="va">n_strains</span><span class="op">)</span></span>
<span><span class="va">observation_times</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind_rows.html" class="external-link">bind_rows</a></span><span class="op">(</span><span class="va">obs1</span>, <span class="va">obs2</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="run-simulation">8. Run simulation<a class="anchor" aria-label="anchor" href="#run-simulation"></a>
</h2>
<p>We now have all the pieces in place to run the full <em>serosim</em>
model. Note that for this example, we will use <em>serosim</em>’s
optional arguments for speeding up the model, exploiting both the
attempted precomputation step and parellelization.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Run the core simulation</span></span>
<span><span class="va">res</span><span class="op">&lt;-</span> <span class="fu"><a href="../reference/runserosim.html">runserosim</a></span><span class="op">(</span></span>
<span>  simulation_settings<span class="op">=</span><span class="va">simulation_settings</span>,</span>
<span>  demography<span class="op">=</span><span class="va">demography</span>,</span>
<span>  observation_times<span class="op">=</span><span class="va">observation_times</span>,</span>
<span>  foe_pars<span class="op">=</span><span class="va">foe_pars</span>,</span>
<span>  biomarker_map<span class="op">=</span><span class="va">antigenic_map</span>,</span>
<span>  model_pars<span class="op">=</span><span class="va">model_pars</span>,</span>
<span>  exposure_model<span class="op">=</span><span class="va">exposure_model_simple_FOE</span>,</span>
<span>  immunity_model<span class="op">=</span><span class="va">immunity_model_ifxn_biomarker_prot</span>,</span>
<span>  antibody_model<span class="op">=</span><span class="va">antibody_model_monophasic_cross_reactivity</span>,</span>
<span>  observation_model<span class="op">=</span><span class="va">observation_model_discrete_noise</span>,</span>
<span>  draw_parameters<span class="op">=</span><span class="va">draw_parameters_random_fx</span>,</span>
<span></span>
<span>  <span class="co">## Other arguments needed</span></span>
<span>  cutoffs<span class="op">=</span><span class="va">cutoffs</span>,</span>
<span>  sensitivity<span class="op">=</span><span class="va">sensitivity</span>,</span>
<span>  specificity<span class="op">=</span><span class="va">specificity</span>,</span>
<span>  VERBOSE<span class="op">=</span><span class="cn">NULL</span>,</span>
<span>  attempt_precomputation <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  parallel<span class="op">=</span><span class="cn">TRUE</span>,</span>
<span>  n_cores<span class="op">=</span><span class="fl">8</span></span>
<span><span class="op">)</span></span>
<span><span class="co">## Note that models and arguments specified earlier in the code can be specified directly within this function.</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="generate-output-plots">9. Generate output plots<a class="anchor" aria-label="anchor" href="#generate-output-plots"></a>
</h2>
<p>Now that the simulation is complete, let’s plot and examine the
simulation outputs. First, let’s look at the full longitudinal kinetics
to one biomarker for a random subset of individuals, marking the timing
of each infection and the exposure ID. The vertical dashed lines show
the timing of infections, colored by their exposure ID.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Plot influenza strain A specific biomarker kinetics and exposure histories for 10 individuals </span></span>
<span><span class="fu"><a href="../reference/plot_subset_individuals_history.html">plot_subset_individuals_history</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">biomarker_states</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">b</span><span class="op">==</span><span class="fl">1</span><span class="op">)</span>, <span class="va">res</span><span class="op">$</span><span class="va">immune_histories_long</span>, subset<span class="op">=</span><span class="fl">10</span>, <span class="va">demography</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: Removed 236 rows containing missing values (`geom_line()`).</span></span></code></pre>
<p><img src="case_study_3_influenza_files/figure-html/unnamed-chunk-12-1.png" width="768"></p>
<p>We can also look at the biomarker kinetics for all biomarker IDs
simultaneously by setting the <code>heatmap</code> argument to
<code>TRUE</code>.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Plot influenza strain A specific biomarker kinetics and exposure histories for 10 individuals </span></span>
<span><span class="fu"><a href="../reference/plot_subset_individuals_history.html">plot_subset_individuals_history</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">biomarker_states</span>, <span class="va">res</span><span class="op">$</span><span class="va">immune_histories_long</span>, subset<span class="op">=</span><span class="fl">10</span>, <span class="va">demography</span>, heatmap<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="case_study_3_influenza_files/figure-html/unnamed-chunk-13-1.png" width="768"></p>
<p>We can see the time-varying exposure probabilities of the different
strains, as well as the individual-level exposure probabilities mediated
by their birth time.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Plot individual probability of exposure for all exposure types.</span></span>
<span><span class="co">## This is the output of the exposure model. Note that this reflects our inputs that the influenza strains were never co-circulating.</span></span>
<span><span class="fu"><a href="../reference/plot_exposure_force.html">plot_exposure_force</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">exposure_force_long</span><span class="op">)</span></span></code></pre></div>
<p><img src="case_study_3_influenza_files/figure-html/unnamed-chunk-14-1.png" width="768"></p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Plot individual successful exposure probabilities for all exposure types</span></span>
<span><span class="co">## This is the output of the exposure model multiplied by the output of the immunity model.</span></span>
<span><span class="co">## In other words, this is the probability of exposure event being successful and inducing an immunological response</span></span>
<span><span class="fu"><a href="../reference/plot_exposure_prob.html">plot_exposure_prob</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">exposure_probabilities_long</span><span class="op">)</span></span></code></pre></div>
<p><img src="case_study_3_influenza_files/figure-html/unnamed-chunk-15-1.png" width="768"></p>
<p>We also track each individual’s history of infections and their true
latent biomarker quantity against each biomarker ID.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Plot individual exposure histories for all exposure types</span></span>
<span><span class="fu"><a href="../reference/plot_immune_histories.html">plot_immune_histories</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">immune_histories_long</span><span class="op">)</span></span></code></pre></div>
<p><img src="case_study_3_influenza_files/figure-html/unnamed-chunk-16-1.png" width="768"></p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Plot biomarket states for all individuals over time</span></span>
<span><span class="fu"><a href="../reference/plot_biomarker_quantity.html">plot_biomarker_quantity</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">biomarker_states</span><span class="op">)</span></span></code></pre></div>
<p><img src="case_study_3_influenza_files/figure-html/unnamed-chunk-17-1.png" width="768"></p>
<p>Finally, we are interested in simulating a realistic serological
survey. The following plots show the sort of observations that we would
obtain from paired samples collected one year apart and tested for HI
titres against each of the 10 strains.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Plot the serosurvey results (observed biomarker quantities)</span></span>
<span><span class="fu"><a href="../reference/plot_obs_biomarkers_one_sample.html">plot_obs_biomarkers_one_sample</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">observed_biomarker_states</span>,add_boxplot<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="case_study_3_influenza_files/figure-html/unnamed-chunk-18-1.png" width="480"></p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Plot the influenza strain A serosurvey results (observed biomarker quantities)</span></span>
<span><span class="fu"><a href="../reference/plot_obs_biomarkers_paired_sample.html">plot_obs_biomarkers_paired_sample</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">observed_biomarker_states</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: Removed 22 rows containing missing values (`geom_line()`).</span></span></code></pre>
<pre><code><span><span class="co">## Warning: Removed 40 rows containing missing values (`geom_point()`).</span></span></code></pre>
<p><img src="case_study_3_influenza_files/figure-html/unnamed-chunk-19-1.png" width="768"></p>
</div>
<div class="section level2">
<h2 id="conclusions">10. Conclusions<a class="anchor" aria-label="anchor" href="#conclusions"></a>
</h2>
<p>In this vignette, we have demonstrated how a complex epidemiological
system representing multi-strain A/H3N2 serology can be implemented
using <em>serosim</em>’s plug-and-play framework. Although this
simulation is not intended to be an exhaustive, realistic representation
of A/H3N2 dynamics, it demonstrates how the multiple layers of the
data-generating process interact to give rise to complex immune
landscapes over time. Alternative models and parameterizations can be
implemented to reflect different pathogens.</p>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Krammer2019" class="csl-entry">
Krammer, Florian. 2019. <span>“The Human Antibody Response to Influenza
a Virus Infection and Vaccination.”</span> <em>Nature Reviews
Immunology</em>, March, 1. <a href="https://doi.org/10.1038/s41577-019-0143-6" class="external-link">https://doi.org/10.1038/s41577-019-0143-6</a>.
</div>
<div id="ref-Kucharski2018" class="csl-entry">
Kucharski, Adam J., Justin Lessler, Derek A. T. Cummings, and Steven
Riley. 2018. <span>“Timescales of Influenza a/H3N2 Antibody
Dynamics.”</span> Edited by Sarah Rowland-Jones. <em>PLOS Biology</em>
16 (August): e2004974. <a href="https://doi.org/10.1371/journal.pbio.2004974" class="external-link">https://doi.org/10.1371/journal.pbio.2004974</a>.
</div>
<div id="ref-Lessler2012" class="csl-entry">
Lessler, Justin, Steven Riley, Jonathan M. Read, Shuying Wang, Huachen
Zhu, Gavin J D Smith, Yi Guan, Chao Qiang Jiang, and Derek A T Cummings.
2012. <span>“Evidence for Antigenic Seniority in Influenza a (H3N2)
Antibody Responses in Southern China.”</span> <em>PLoS Pathogens</em> 8
(July): e1002802. <a href="https://doi.org/10.1371/journal.ppat.1002802" class="external-link">https://doi.org/10.1371/journal.ppat.1002802</a>.
</div>
<div id="ref-Smith2004" class="csl-entry">
Smith, Derek J, Alan S Lapedes, Jan C de Jong, Theo M Bestebroer, Guus F
Rimmelzwaan, Albert D M E Osterhaus, and Ron A M Fouchier. 2004.
<span>“Mapping the Antigenic and Genetic Evolution of Influenza
Virus.”</span> <em>Science</em> 305 (July): 371–76. <a href="https://doi.org/10.1126/science.1097211" class="external-link">https://doi.org/10.1126/science.1097211</a>.
</div>
<div id="ref-Yang2020" class="csl-entry">
Yang, Bingyi, Justin Lessler, Huachen Zhu, Chao Qiang Jiang, Jonathan M.
Read, James A. Hay, Kin On Kwok, et al. 2020. <span>“Life Course
Exposures Continually Shape Antibody Profiles and Risk of Seroconversion
to Influenza.”</span> Edited by Colin A. Russell. <em>PLOS
Pathogens</em> 16 (July): e1008635. <a href="https://doi.org/10.1371/journal.ppat.1008635" class="external-link">https://doi.org/10.1371/journal.ppat.1008635</a>.
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Arthur Menezes, James Hay.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
